File: Source/Scriptit-Core.swiftpm/Resources/Core/Scriptit/Hud/hud.js
///////////////////////////////////////////////////////////
// HUD MODULE
///////////////////////////////////////////////////////////

/** Singleton class representing the main hud object. */
class HudManager
{
  #errors;
  static #instance = null;

  /** Creates the hud object. **/
  constructor() 
  {
    this.#errors = 
    {
      messageTypeError: 'Hud Manager Error: Expected type string for message. ',
      singleInstanceError: 'Hud Manager Error: Only one HudManager object can exist at a time.',
      timoeutTypeError: 'Hud Manager Error: Expected type number for timeout. ',
    };

    if(HudManager.#instance) console.error(this.#errors.singleInstanceError);
    else HudManager.#instance = this;
  }

  /** Static method to return a new HudManager instance. */
  static getInstance() 
  {
    return new HudManager();
  }

  /**
 * Internal helper to safely send commands to the native HUD.
 * @param {string} command - The HUD command (added, failed, loading, succeed)
 * @param {Object} [options={}] - Optional { message, timeout } object.
 */
  #send(command, options = {}) 
  {
    let payload = { command };
    if(options && typechecker.check({ type: 'object', value: options })) 
    {
      if(options.message !== undefined) 
      {
        if(!typechecker.check({ type: 'string', value: options.message })) 
        {
          console.error(this.#errors.messageTypeError);
          return;
        }
        payload.message = options.message;
      }
      
      options.timeout = 1500;
      
      if(options.timeout !== undefined) 
      {
        if(!typechecker.check({ type: 'number', value: options.timeout })) 
        {
          console.error(this.#errors.timeoutTypeError);
          return;
        }
  
        let convertedTimeout = Math.round((options.timeout / 1000) * 10) / 10;
        payload.timeout = convertedTimeout;
      }
    }
  
    window.webkit?.messageHandlers?.hudMessageManager?.postMessage(payload);
  }


  /**
   * Public method to show "added" HUD.
   * @param {Object} [options={}] - Optional { message, delay } object.
   */
  added(options = {}) 
  { 
    this.#send('added', options); 
  }
  
   
  /** Public method to dismiss the HUD manually. */
  dismiss()
  {
    window.webkit?.messageHandlers?.hudMessageManager?.postMessage({ command: 'dismiss' });
  }

  /**
   * Public method to show "failed" HUD.
   * @param {Object} [options={}] - Optional { message, delay } object.
   */
  failed(options = {}) 
  { 
    this.#send('failed', options); 
  }

  /**
   * Public method to show "loading" HUD.
   * @param {Object} [options={}] - Optional { message, delay } object.
   */
  loading(options = {}) 
  { 
    this.#send('loading', options); 
  }

  /**
   * Public method to show "success" HUD.
   * @param {Object} [options={}] - Optional { message, delay } object.
   */
  succeed(options = {}) 
  { 
    this.#send('succeed', options); 
  }
}

///////////////////////////////////////////////////////////

globalThis.hud = HudManager.getInstance();

///////////////////////////////////////////////////////////

File: Source/Scriptit-Core.swiftpm/iOS-Core/Hud/Hud.swift
//=============================================//

import UIKit
import WebKit
import ProgressHUD

//=============================================//

/** Class that manages messages for the hud module. */
class HudMessageManager: JavascriptMessageManager
{
  enum Errors: String
  {
    case invalidMessageBody = "HUD Error: Invalid message body."
    case commandNotProvided = "HUD Error: Command not provided."
    case invalidCommand = "HUD Error: Invalid command."
    case invalidMessage = "HUD Error: Invalid message."
    case invalidTimeout = "HUD Error: Invalid timeout value."
  }
  
  let errors = Errors.self;

  /**
   * Entry point for all JavaScript messages targeting the HUD module.
   *
   * This method validates the incoming message payload, extracts the
   * required command and optional parameters, and routes execution
   * to the appropriate HUD action.
   *
   * Invalid input or unsupported commands are logged using standardized
   * error messages and do not crash the application.
   *
   * Expected JavaScript input (dict):
   * - command (String): HUD command ("added", "failed", "loading", "succeed", "dismiss")
   * - message (String, optional): HUD display text
   * - timeout (Double, optional): Delay before auto-dismiss
   *
   * @param message The WKScriptMessage sent from JavaScript.
   * @param webView The WKWebView instance associated with the message.
   */
  func handleMessage(_ message: WKScriptMessage, webView: WKWebView)
  {
    if let windowScene = UIApplication.shared.connectedScenes
      .compactMap({ $0 as? UIWindowScene })
      .first,
       let window = windowScene.windows.first(where: { $0.isKeyWindow })
    {
      window.overrideUserInterfaceStyle = .light
    }
    
    guard let dict = message.body as? [String: Any] else
    {
      let error = self.errors.invalidMessageBody.rawValue;
      print(error);
      return;
    }

    guard let command = dict["command"] as? String else
    {
      let error = self.errors.commandNotProvided.rawValue;
      print(error);
      return;
    }

    let hudMessage = dict["message"] as? String;
    let timeoutValue = dict["timeout"];

    var delay: Double = 0;
    if let timeout = timeoutValue
    {
      guard let parsedDelay = timeout as? Double else
      {
        let error = self.errors.invalidTimeout.rawValue;
        print(error);
        return;
      }
      delay = parsedDelay;
    }

    switch command
    {
      case "added":
        self.handleAdded(message: hudMessage, delay: delay)
      case "failed":
        self.handleFailed(message: hudMessage, delay: delay)
      case "loading":
        self.handleLoading(message: hudMessage)
      case "succeed":
        self.handleSucceed(message: hudMessage, delay: delay)
      case "dismiss":
        ProgressHUD.dismiss()
      default:
        let error = self.errors.invalidCommand.rawValue + " (\(command)).";
        print(error);
    }
  }

  /**
   * Displays a success-style HUD indicating an item was added.
   *
   * If a message is provided, it is displayed to the user.
   * If a delay is greater than zero, the HUD auto-dismisses after the delay.
   *
   * @param message Optional text to display in the HUD.
   * @param delay Optional delay before dismissal.
   */
  func handleAdded(message: String?, delay: Double)
  {
    if let msg = message { delay > 0 ? ProgressHUD.added(msg, delay: delay) : ProgressHUD.added(msg); }
    else { delay > 0 ? ProgressHUD.added(delay: delay) : ProgressHUD.added(); }
  }

  /**
   * Displays an error-style HUD indicating a failure occurred.
   *
   * If a message is provided, it is displayed to the user.
   * If a delay is greater than zero, the HUD auto-dismisses after the delay.
   *
   * @param message Optional text to display in the HUD.
   * @param delay Optional delay before dismissal.
   */
  func handleFailed(message: String?, delay: Double)
  {
    if let msg = message { delay > 0 ? ProgressHUD.failed(msg, delay: delay) : ProgressHUD.failed(msg); }
    else { delay > 0 ? ProgressHUD.failed(delay: delay) : ProgressHUD.failed(); }
  }

  /**
   * Displays a loading or indeterminate-progress HUD.
   *
   * If a message is provided, it is displayed alongside the animation.
   *
   * @param message Optional loading text to display.
   */
  func handleLoading(message: String?)
  {
    if let msg = message { ProgressHUD.animate(msg); }
    else { ProgressHUD.animate(); }
  }

  /**
   * Displays a success-style HUD indicating a successful operation.
   *
   * If a message is provided, it is displayed to the user.
   * If a delay is greater than zero, the HUD auto-dismisses after the delay.
   *
   * @param message Optional text to display in the HUD.
   * @param delay Optional delay before dismissal.
   */
  func handleSucceed(message: String?, delay: Double)
  {
    if let msg = message { delay > 0 ? ProgressHUD.succeed(msg, delay: delay) : ProgressHUD.succeed(msg); }
    else { delay > 0 ? ProgressHUD.succeed(delay: delay) : ProgressHUD.succeed(); }
  }
}

//=======================================================//